src/Gui/WorkbenchSelector.cpp

```diff
@@ -101,156 +104,307 @@ void WorkbenchComboBox::refreshList(QList<QAction*> actionList)
-        setShape(area == Qt::LeftToolBarArea ? QTabBar::RoundedWest : QTabBar::RoundedEast);
-        hGrp->SetASCII("TabBarOrientation", area == Qt::LeftToolBarArea ? "West" : "East");
+    if (temporaryWorkbenchAction != nullptr) {
+        setTemporaryWorkbenchTab(temporaryWorkbenchAction);
+    }
+
+    buildPrefMenu();
+    adjustSize();
+}
+
+int WorkbenchTabWidget::addWorkbenchTab(QAction* action, int tabIndex)
+{
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Workbenches");
+    auto itemStyle = static_cast<WorkbenchItemStyle>(hGrp->GetInt("WorkbenchSelectorItem", 0));
+
+    // if tabIndex is negative we assume that tab must be placed at the end of tabBar (default behavior)
+    if (tabIndex < 0) {
+        tabIndex = tabBar->count();
+    }
+
+    // for the maps we consider order in which tabs have been added
+    // that's why here we use tabBar->count() instead of tabIndex
+    actionToTabIndex[action] = tabBar->count();
+    tabIndexToAction[tabBar->count()] = action;
+
+    QIcon icon = action->icon();
+    if (icon.isNull() || itemStyle == WorkbenchItemStyle::TextOnly) {
+        tabBar->insertTab(tabIndex, action->text());
+    }
+    else if (itemStyle == IconOnly) {
+        tabBar->insertTab(tabIndex, icon, {}); // empty string to ensure only icon is displayed
     }
     else {
-        setShape(area == Qt::TopToolBarArea ? QTabBar::RoundedNorth : QTabBar::RoundedSouth);
-        hGrp->SetASCII("TabBarOrientation", area == Qt::TopToolBarArea ? "North" : "South");
+        tabBar->insertTab(tabIndex, icon, action->text());
+    }
+
+    tabBar->setTabToolTip(tabIndex, action->toolTip());
+
+    if (action->isChecked()) {
+        tabBar->setCurrentIndex(tabIndex);
+    }
+
+    return tabIndex;
+}
+
+void WorkbenchTabWidget::setToolBarArea(Gui::ToolBarArea area)
+{
+    switch (area) {
+        case Gui::ToolBarArea::LeftToolBarArea:
+        case Gui::ToolBarArea::RightToolBarArea: {
+            setDirection(Qt::LeftToRight);
+            layout->setDirection(direction() == Qt::LeftToRight ? QBoxLayout::TopToBottom : QBoxLayout::BottomToTop);
+            tabBar->setShape(area == Gui::ToolBarArea::LeftToolBarArea ? QTabBar::RoundedWest : QTabBar::RoundedEast);
+            break;
+        }
+
+        case Gui::ToolBarArea::TopToolBarArea:
+        case Gui::ToolBarArea::BottomToolBarArea:
+        case Gui::ToolBarArea::LeftMenuToolBarArea:
+        case Gui::ToolBarArea::RightMenuToolBarArea:
+        case Gui::ToolBarArea::StatusBarToolBarArea: {
+            bool isTop =
+                area == Gui::ToolBarArea::TopToolBarArea ||
+                area == Gui::ToolBarArea::LeftMenuToolBarArea ||
+                area == Gui::ToolBarArea::RightMenuToolBarArea;
+
+            bool isRightAligned =
+                area == Gui::ToolBarArea::RightMenuToolBarArea ||
+                area == Gui::ToolBarArea::StatusBarToolBarArea;
```

Comment:
shouldn't this also take into account if the locale is set to a rtl language?

Reply:
Hard to say for me, It depends on how Qt handles RTL languages and honestly I'm unsure on how it should behave. It may require more logic to handle this properly, so if you see any particular problem please create an issue and I'll fix it soon

Reply:
I don't use a RTL language but I think it's important to pay attention to these details too. I'll try to test one next week and create an issue if there is one

Reply:
Sure, agree with that but it's hard for person who does not work with RTL languages to figure out how it should behave in details, as I presume it is not as simple as basically mirror everything. I'll also play with it a bit to check if it works properly. Thanks for pointing that out!

Reply:
@shaise as person who uses an RTL language, could you please weigh in?